flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nlp/Word2Vec.scala

```diff
@@ -0,0 +1,243 @@
+  *
+  * - [[org.apache.flink.ml.nlp.Word2Vec.VectorSize]]
+  * sets the length of each learned vector (Default value: '''100''')
+  *
+  * - [[org.apache.flink.ml.nlp.Word2Vec.LearningRate]]
+  * sets the rate of descent during backpropagation - this value decays linearly with
+  * individual training sets, determined by BatchSize (Default value: '''0.015''')
+  *
+  * - [[org.apache.flink.ml.nlp.Word2Vec.BatchSize]]
+  * sets the batch size of training sets - the input DataSet will be batched into
+  * groups of this size for learning (Default value: '''1000''')
+  *
+  * - [[org.apache.flink.ml.nlp.Word2Vec.Seed]]
+  * sets the seed for generating random vectors at initial weighting DataSet creation
+  * (Default value: '''Some(scala.util.Random.nextLong)''')
+  */
+class Word2Vec extends Transformer[Word2Vec] {
+  import Word2Vec._
+
+  private [nlp] var wordVectors:
+    Option[DataSet[HSMWeightMatrix[String]]] = None
+
+  def setIterations(iterations: Int): this.type = {
+    parameters.add(Iterations, iterations)
+    this
+  }
+
+  def setTargetCount(targetCount: Int): this.type = {
+    parameters.add(TargetCount, targetCount)
+    this
+  }
+
+  def setVectorSize(vectorSize: Int): this.type = {
+    parameters.add(VectorSize, vectorSize)
+    this
+  }
+
+  def setLearningRate(learningRate: Double): this.type = {
+    parameters.add(LearningRate, learningRate)
+    this
+  }
+
+  def setWindowSize(windowSize: Int): this.type = {
+    parameters.add(WindowSize, windowSize)
+    this
+  }
+
+  def setBatchSize(batchSize: Int): this.type = {
+    parameters.add(BatchSize, batchSize)
+    this
+  }
+
+  def setSeed(seed: Long): this.type = {
+    parameters.add(Seed, seed)
+    this
+  }
+
+}
+
+object Word2Vec {
+  case object Iterations extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object TargetCount extends Parameter[Int] {
+    val defaultValue = Some(5)
+  }
+
+  case object VectorSize extends Parameter[Int] {
+    val defaultValue = Some(100)
+  }
+
+  case object LearningRate extends Parameter[Double] {
+    val defaultValue = Some(0.015)
+  }
+
+  case object WindowSize extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object BatchSize extends Parameter[Int] {
+    val defaultValue = Some(1000)
+  }
+
+  case object Seed extends Parameter[Long] {
+    val defaultValue = Some(scala.util.Random.nextLong)
+  }
+
+  def apply(): Word2Vec = {
+    new Word2Vec()
+  }
+
+  /** [[FitOperation]] which builds initial vocabulary for Word2Vec context embedding
+    *
+    * @tparam T Subtype of Iterable[String]
+    * @return
+    */
+  implicit def learnWordVectors[T <: Iterable[String]] = {
+    new FitOperation[Word2Vec, T] {
+      override def fit(
+        instance: Word2Vec,
+        fitParameters: ParameterMap,
+        input: DataSet[T])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+        
+        val skipGrams = input
```

Comment:
By that I mean, that your code, will start to perform many initialization steps for trivial input of size =0, for example it will go to initialization of ContextEmbedder to the method .createInitialWeightsDS(instance.wordVectors, skipGrams) and further, but this is not necessary for trivial case. Please perform the check of trivial situations.